### PR TITLE
[NPM-2924] adds netlink error logs and tags

### PR DESCRIPTION
**Pull request by CordeliaLopez**
### What does this PR do?

Logs a warning when we find netlink errors and adds the errno as tag for error counters.
 
### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/route_cache.go [hunk 1]
```diff
@@ -275,13 +275,20 @@ func (n *netlinkRouter) Route(source, dest util.Address, netns uint32) (Route, b
...
 	if err != nil {
-		routeCacheTelemetry.netlinkErrors.Inc()
+		var tag string
 		if iifIndex > 0 {
-			if errno, ok := err.(syscall.Errno); ok && (errno == syscall.EINVAL || errno == syscall.ENODEV) {
-				// invalidate interface cache entry as this may have been the cause of the netlink error
-				n.removeInterface(source, netns)
+			errno, ok := err.(syscall.Errno)
+			if ok {
+				tag = errno.Error()
```
Comment (hmahmood) [inline pkg/network/route_cache.go:282]: Does this give the description or the error code, e.g. `EINVAL`? We want the latter.

File: pkg/network/route_cache.go [hunk 2]
```diff
@@ -275,13 +275,20 @@ func (n *netlinkRouter) Route(source, dest util.Address, netns uint32) (Route, b
...
+				if errno == syscall.EINVAL || errno == syscall.ENODEV {
+					// invalidate interface cache entry as this may have been the cause of the netlink error
+					n.removeInterface(source, netns)
+				}
 			}
 		}
+		log.Warnf("Netlink Error: %s", err)
```
Comment (hmahmood) [inline pkg/network/route_cache.go:289]: We should put more context in the message; just "netlink error" is too generic.

File: pkg/network/route_cache.go [hunk 3]
```diff
@@ -315,7 +322,13 @@ func (n *netlinkRouter) getInterface(srcAddress util.Address, srcIP net.IP, netn
 	routeCacheTelemetry.netlinkLookups.Inc()
 	routes, err := n.nlHandle.RouteGet(srcIP)
 	if err != nil {
-		routeCacheTelemetry.netlinkErrors.Inc()
+		var tag string
```
Comment (hmahmood) [inline pkg/network/route_cache.go:325]: May want to add a helper for errno to string tag.

File: pkg/network/route_cache.go [hunk 4]
```diff
@@ -315,7 +322,13 @@ func (n *netlinkRouter) getInterface(srcAddress util.Address, srcIP net.IP, netn
...
+		errno, ok := err.(syscall.Errno)
+		if ok {
+			tag = errno.Error()
+		}
+		log.Warnf("Netlink Error: %s", err)
```
Comment (hmahmood) [inline pkg/network/route_cache.go:330]: Ditto here ... need more details in the log message.

File: pkg/network/route_cache.go (hunk 2 above)
Comment (CordeliaLopez) [reply pkg/network/route_cache.go]: lmk if the new message is enough

File: pkg/network/route_cache.go (hunk 1 above)
Comment (CordeliaLopez) [reply pkg/network/route_cache.go]: fixed to give the code

File: pkg/network/route_cache.go [hunk 5]
```diff
@@ -275,14 +275,19 @@ func (n *netlinkRouter) Route(source, dest util.Address, netns uint32) (Route, b
...
+			errno, ok := err.(syscall.Errno)
+			if ok {
+				netlinkErrIncWithTag(errno)
+				if errno == syscall.EINVAL || errno == syscall.ENODEV {
+					// invalidate interface cache entry as this may have been the cause of the netlink error
+					n.removeInterface(source, netns)
+				}
 			}
 		}
+		log.Debug("Error getting route via netlink from %s: %s", string(dstIP), err)
```
Comment (hmahmood) [inline pkg/network/route_cache.go:288]: `Debugf`

File: pkg/network/route_cache.go [hunk 6]
```diff
@@ -275,14 +275,19 @@ func (n *netlinkRouter) Route(source, dest util.Address, netns uint32) (Route, b
...
 
 	if err != nil {
-		routeCacheTelemetry.netlinkErrors.Inc()
 		if iifIndex > 0 {
-			if errno, ok := err.(syscall.Errno); ok && (errno == syscall.EINVAL || errno == syscall.ENODEV) {
-				// invalidate interface cache entry as this may have been the cause of the netlink error
-				n.removeInterface(source, netns)
+			errno, ok := err.(syscall.Errno)
+			if ok {
+				netlinkErrIncWithTag(errno)
```
Comment (hmahmood) [inline pkg/network/route_cache.go:281]: I don't think should only be incremented if err is an errno.

File: pkg/network/route_cache.go [hunk 7]
```diff
@@ -315,9 +320,14 @@ func (n *netlinkRouter) getInterface(srcAddress util.Address, srcIP net.IP, netn
 	routeCacheTelemetry.netlinkLookups.Inc()
 	routes, err := n.nlHandle.RouteGet(srcIP)
 	if err != nil {
-		routeCacheTelemetry.netlinkErrors.Inc()
+		errno, ok := err.(syscall.Errno)
+		if ok {
+			netlinkErrIncWithTag(errno)
```
Comment (hmahmood) [inline pkg/network/route_cache.go:325]: Ditto here

File: pkg/network/route_cache.go [hunk 8]
```diff
@@ -275,14 +275,19 @@ func (n *netlinkRouter) Route(source, dest util.Address, netns uint32) (Route, b
...
+				netlinkErrIncWithTag(errno)
+				if errno == syscall.EINVAL || errno == syscall.ENODEV {
+					// invalidate interface cache entry as this may have been the cause of the netlink error
+					n.removeInterface(source, netns)
+				}
 			}
 		}
+		log.Debugf("Error getting route via netlink from %s: %s", dstIP, err)
 	} else if len(routes) != 1 {
+		log.Debugf("Failed to get route from cache for %s", dstIP)
```
Comment (hmahmood) [inline pkg/network/route_cache.go:290]: Can we make this more specific, like `Did not exactly one route for %s, got %d routes`?

File: pkg/network/route_cache.go [hunk 9]
```diff
@@ -349,3 +359,11 @@ func (n *netlinkRouter) getInterface(srcAddress util.Address, srcIP net.IP, netn
 	routeCacheTelemetry.ifCacheSize.Inc()
 	return iff
 }
+
+func netlinkErrIncWithTag(err syscall.Errno) {
```
Comment (hmahmood) [inline pkg/network/route_cache.go:363]: You can likely change `err` to a `error`, and then try type casting to `syscall.Errno`. And then based on that, tag or not.

File: pkg/network/route_cache.go [hunk 10]
```diff
@@ -275,14 +275,19 @@ func (n *netlinkRouter) Route(source, dest util.Address, netns uint32) (Route, b
...
+			errno, ok := err.(syscall.Errno)
+			if ok {
+				netlinkErrIncWithTag(errno)
+				if errno == syscall.EINVAL || errno == syscall.ENODEV {
+					// invalidate interface cache entry as this may have been the cause of the netlink error
+					n.removeInterface(source, netns)
+				}
 			}
 		}
+		log.Debugf("Error getting route via netlink from %s: %s", dstIP, err)
```
Comment (hmahmood) [inline pkg/network/route_cache.go:288]: `from %s` -> `for destination IP %s`

File: pkg/network/route_cache.go [hunk 11]
```diff
@@ -349,3 +355,18 @@ func (n *netlinkRouter) getInterface(srcAddress util.Address, srcIP net.IP, netn
 	routeCacheTelemetry.ifCacheSize.Inc()
 	return iff
 }
+
+func incWithTag(counter telemetry.Counter, err error) (errno syscall.Errno, ok bool) {
```
Comment (hmahmood) [inline pkg/network/route_cache.go:359]: `CounterIncWithTag`?

File: pkg/network/route_cache.go [hunk 12]
```diff
@@ -315,16 +319,18 @@ func (n *netlinkRouter) getInterface(srcAddress util.Address, srcIP net.IP, netn
...
 	} else if len(routes) != 1 {
+		log.Debugf("Did not get exactly one route for %s, got %d routes", srcIP, len(routes))
 		routeCacheTelemetry.netlinkMisses.Inc()
 		return nil
 	}
 
 	ifr, err := unix.NewIfreq("")
 	if err != nil {
-		routeCacheTelemetry.ifCacheErrors.Inc()
+		_, _ = incWithTag(routeCacheTelemetry.ifCacheErrors, err)
```
Comment (hmahmood) [inline pkg/network/route_cache.go:333]: Did you add the tag?

Comment (CordeliaLopez) [reply pkg/network/route_cache.go:333]: ah I knew I had missed that! Thanks

File: pkg/network/route_cache.go [hunk 13]
```diff
@@ -275,14 +275,18 @@ func (n *netlinkRouter) Route(source, dest util.Address, netns uint32) (Route, b
 		})
 
 	if err != nil {
-		routeCacheTelemetry.netlinkErrors.Inc()
 		if iifIndex > 0 {
-			if errno, ok := err.(syscall.Errno); ok && (errno == syscall.EINVAL || errno == syscall.ENODEV) {
-				// invalidate interface cache entry as this may have been the cause of the netlink error
-				n.removeInterface(source, netns)
+			errno, ok := IncWithTag(routeCacheTelemetry.netlinkErrors, err)
```
Comment (hmahmood) [inline pkg/network/route_cache.go:279]: `CounterIncWithTag`?

File: pkg/network/route_cache.go [hunk 14]
```diff
@@ -349,3 +355,19 @@ func (n *netlinkRouter) getInterface(srcAddress util.Address, srcIP net.IP, netn
 	routeCacheTelemetry.ifCacheSize.Inc()
 	return iff
 }
+
+// Adds error tag and increments counter
+func CounterIncWithTag(counter telemetry.Counter, err error) (errno syscall.Errno, ok bool) {
+	errno, ok = err.(syscall.Errno)
```
Comment (hmahmood) [inline pkg/network/route_cache.go:361]: You can make this more concise:
```go
if errno, ok = err.(syscall.Errno); ok {
  	if tag := unix.ErrnoName(errno); tag != "" {
		counter.Inc(tag)
		return
	}
}

counter.Inc()
return
```

Comment (CordeliaLopez) [reply pkg/network/route_cache.go:361]: I think something's weird with my editor... I had already changed that

File: pkg/network/route_cache.go [hunk 15]
```diff
@@ -275,14 +275,18 @@ func (n *netlinkRouter) Route(source, dest util.Address, netns uint32) (Route, b
 		})
 
 	if err != nil {
-		routeCacheTelemetry.netlinkErrors.Inc()
 		if iifIndex > 0 {
-			if errno, ok := err.(syscall.Errno); ok && (errno == syscall.EINVAL || errno == syscall.ENODEV) {
-				// invalidate interface cache entry as this may have been the cause of the netlink error
-				n.removeInterface(source, netns)
+			errno, ok := CounterIncWithTag(routeCacheTelemetry.netlinkErrors, err)
+			if ok {
```
Comment (mbakht) [inline pkg/network/route_cache.go:280]: Do we need to split the condition check into multiple `if`s?

File: pkg/network/route_cache.go [hunk 16]
```diff
@@ -349,3 +353,15 @@ func (n *netlinkRouter) getInterface(srcAddress util.Address, srcIP net.IP, netn
 	routeCacheTelemetry.ifCacheSize.Inc()
 	return iff
 }
+
+// CounterIncWithTag adds error tag and increments counter
+func CounterIncWithTag(counter telemetry.Counter, err error) (errno syscall.Errno, ok bool) {
```
Comment (hmahmood) [inline pkg/network/route_cache.go:358]: Does this need to be exported?

File: pkg/network/route_cache.go [hunk 17]
```diff
@@ -275,14 +275,16 @@ func (n *netlinkRouter) Route(source, dest util.Address, netns uint32) (Route, b
 		})
 
 	if err != nil {
-		routeCacheTelemetry.netlinkErrors.Inc()
 		if iifIndex > 0 {
-			if errno, ok := err.(syscall.Errno); ok && (errno == syscall.EINVAL || errno == syscall.ENODEV) {
+			errno, ok := CounterIncWithTag(routeCacheTelemetry.netlinkErrors, err)
```
Comment (hmahmood) [inline pkg/network/route_cache.go:279]: This should be outside the `if`?

File: pkg/network/route_cache.go (hunk 16 above)
Comment (CordeliaLopez) [reply pkg/network/route_cache.go:358]: No it does not... I misinterpreted one of your previous comments

**Review by hmahmood — APPROVED** [review]

**Pull request merged**